### PR TITLE
Upgrade pulsar version to 2.1.0-incubating #75

### PR DESCRIPTION
**Pull request by FNegrello**
Upgrading pulsar driver with the latest version of pulsar client 2.1.0-incubating.
This version is backward compatible with pulsar 1.22.0

### CODE REVIEW

File: driver-pulsar/src/main/java/io/openmessaging/benchmark/driver/pulsar/PulsarBenchmarkDriver.java [hunk 1]
```diff
@@ -63,63 +61,61 @@
...
-                    adminClient.properties().createProperty(property,
-                            new PropertyAdmin(Collections.emptyList(), Sets.newHashSet(cluster)));
+                    adminClient.tenants().createTenant(tenant,
+                                    new TenantInfo(Collections.emptySet(), Sets.newHashSet(cluster)));
                 } catch (ConflictException e) {
                     // Ignore. This can happen when multiple workers are initializing at the same time
                 }
             }
-            log.info("Created Pulsar property {} with allowed cluster {}", property, cluster);
+            log.info("Created Pulsar property {} with allowed cluster {}", tenant, cluster);
```
Comment (merlimat) [inline driver-pulsar/src/main/java/io/openmessaging/benchmark/driver/pulsar/PulsarBenchmarkDriver.java]: Can you also change the log message into `Created Pulsar tenant ...` ?

File: driver-pulsar/src/main/java/io/openmessaging/benchmark/driver/pulsar/PulsarBenchmarkProducer.java [hunk 2]
```diff
@@ -41,12 +41,13 @@ public void close() throws Exception {
...
     public CompletableFuture<Void> sendAsync(Optional<String> key, byte[] payload) {
-        MessageBuilder msgBuilder = MessageBuilder.create().setContent(payload);
+        TypedMessageBuilder<byte[]> msgBuilder = producer.newMessage().value(payload);
         if (key.isPresent()) {
-            msgBuilder.setKey(key.get());
+            msgBuilder.key(key.get());
         }
 
-        return producer.sendAsync(msgBuilder.build()).thenApply(msgId -> null);
+        //return producer.sendAsync(producer).thenApply(msgId -> null);
```
Comment (merlimat) [inline driver-pulsar/src/main/java/io/openmessaging/benchmark/driver/pulsar/PulsarBenchmarkProducer.java]: remove commented line

File: driver-pulsar/src/main/java/io/openmessaging/benchmark/driver/pulsar/PulsarBenchmarkDriver.java [hunk 3]
```diff
@@ -63,63 +61,61 @@
...
-                TimeUnit.MILLISECONDS);
-        producerConfiguration.setBlockIfQueueFull(config.producer.blockIfQueueFull);
-        producerConfiguration.setMaxPendingMessages(config.producer.pendingQueueSize);
-        log.info("Set producer configuration: {}", writer.writeValueAsString(producerConfiguration));
+        producerBuilder = client.newProducer().enableBatching(config.producer.batchingEnabled)
+                        .batchingMaxPublishDelay(config.producer.batchingMaxPublishDelayMs, TimeUnit.MILLISECONDS)
+                        .blockIfQueueFull(config.producer.blockIfQueueFull)
+                        .maxPendingMessages(config.producer.pendingQueueSize);
+
+        //log.info("Set producer configuration: {}", writer.writeValueAsString(producerBuilder));
```
Comment (merlimat) [inline driver-pulsar/src/main/java/io/openmessaging/benchmark/driver/pulsar/PulsarBenchmarkDriver.java]: Remove commented log line

File: driver-pulsar/src/main/java/io/openmessaging/benchmark/driver/pulsar/PulsarBenchmarkDriver.java [hunk 4]
```diff
@@ -63,63 +61,61 @@
...
+                        .serviceUrl(config.client.serviceUrl).build();
 
-        ClientConfiguration clientConfiguration = new ClientConfiguration();
-        clientConfiguration.setIoThreads(config.client.ioThreads);
-        clientConfiguration.setConnectionsPerBroker(config.client.connectionsPerBroker);
 
-        // Disable internal stats since we're already collecting in the framework
-        clientConfiguration.setStatsInterval(0, TimeUnit.SECONDS);
-        log.info("Set Pulsar client configuration: {}", writer.writeValueAsString(clientConfiguration));
+        //log.info("Set Pulsar client configuration: {}", writer.writeValueAsString(client.));
```
Comment (merlimat) [inline driver-pulsar/src/main/java/io/openmessaging/benchmark/driver/pulsar/PulsarBenchmarkDriver.java]: Remove commented log line. There's currently no good way to get a printout of the builder settings. We'll figure something out to have some proper way to add that in the interface.

**Comment by vongosling** [comment]
@jiazhai @merlimat @sijie IMO, it's time to separate the benchmark API from the project. I would like to call for discussion in the next week, hoping to see you :-)

**Comment by sijie** [comment]
> IMO, it's time to separate the benchmark API from the project. I would like to call for discussion in the next week, hoping to see you :-)

@vongosling : the driver-api is well defined under this module - https://github.com/openmessaging/openmessaging-benchmark/tree/master/driver-api

different messaging implementations can implement its own driver already. not sure whether it is worth moving it out, since it is relatively small piece of code.

**Pull request merged**